### ISSUE TITLE
Batch Job to Expire and Reactivate DS & EX Permits with LC Expiration (SS) (on PROD)

### ISSUE THREAD

**Issue by TracyLinder** [cityofaustin/atd-data-tech#10849]
**Expiration Batch Acceptance Criteria**
- System should run a daily batch that checks for LC folders that have changed status from 'Active' to  'Expired' in the last 24 hours. 
- System should then check for the associated DS & EX permits of the 'Right-of-Way Contractor' people record on the folder-people tab of the LC. 
- If there are any DS or EX folders for the People record that have a status of 'Active' and where the People record is associated as 'Right of Way Contractor', system should change the folder status of these folders to 'Expired - License'
- System should also send an email to 'Right-Of-Way Contractor' People record letting them know permits have expired. Text for email below.

**Subject: ROW License and Permit Expiration Notice**

Dear ROW License Holder,

This email is to advise you that your Right of Way Contractor License [2022-12345678 LC] expired on [LC folder Expiry date].

All Driveway/Sidewalk and Excavation permits associated with your Right of Way Contractor License have expired on [LC folder Expiry date] until the license is renewed. A list of those permits is provided below. To renew your Right of Way Contractor License, look for the "How is a Right of Way Contractor License renewed?" Section at this URL - https://austintexas.gov/page/right-way-contractors

Your DS/EX Permits That Will Expire:

- [Folder Number of DS or EX folder, for example, 2022-012345 DS] - [Folder Name]
    - 2023-12345678-DS - New Driveway & Sidewalk
    
To update your insurance, bond or license first view the documentation requirements on the Right of Way Contractor webpage {https://www.austintexas.gov/page/right-way-contractors}, and then submit your updated documentation to our SmartSheet {https://app.smartsheet.com/b/form/cad1e53c88224399979f40df80ef6270}.

**Re-activation process Acceptance Criteria**
- When user makes attempt Renew License in the Renew License Process (58025) in the LC folder, system should then check the People record of LC folders that have a type of 'Right-Of-Way Contractor' to see if they are part of any DS or EX folders 
- If there are any DS or EX folders for the People record that have a status of 'Expired-License' AND have an expiry date in the future, system should change the folder status of these folders to 'Active' 
- System should also send an email to 'Right-Of-Way Contractor' People record letting them know their license has been reactivated. Text for email below.

**Subject: ROW License and Permit Renewal Notice**

Dear ROW License Holder,

This email is to advise you that your Right of Way Contractor License [2022-12345678 LC]  has been reactivated.

All Driveway/Sidewalk and Excavation permits associated with your Right of Way Contractor License have been reactivated. A list of those permits is provided below. Please note that any permits that are past their expiry date will not be reactivated. 

Your DS/EX Permits Reactivated:

- [Folder Number of DS or EX folder, for example, 2022-012345 DS] - [Folder Name]
    - Example: 2023-12345678 DS - New Driveway & Sidewalk

**Modification to DS&EX expiration batches** 
- The batch that expires DS and EX folders should also expire folders in 'Expired-License' status provided the expiry date has lapsed. 
-  Expiration of DS and EX folders happens in batch jobs 38 and 179

**Note** - The status 'Expired-License' does not exist currently. It needs to be newly created. 
**Note** - As an extension to this feature, it was also suggested that the insurance dates can be considered to expire DS and EX folders. This may be addressed as a separate ticket after this one is implemented.

**Comment by TracyLinder** [comment]
### Notes
- [ ] Create new issue to create new valid folder status **Expired - License**

**Comment by TracyLinder** [comment]
**Update**
- Got team feedback
- @vschandramouli will update AC
- Team needs to review updated AC before this goes On Deck

**Comment by TracyLinder** [comment]
### Update

- More team feedback: The external customer (ROW Contractor) will need to be notified when license expires w/ permits, and when the permits a re-activated. Might need to include language about permits that expired during the suspension will not be re-activated. 
FYI @vschandramouli 
- Reference content in #4070

**Comment by vschandramouli** [comment]
Updated AC, ready for Team review/Backlog.

**Comment by TracyLinder** [comment]
### Review w/ ROW

- Assign to Subha
- DTS team needs to meet internally about the way we'll approach this build **before** we begin building
- Turn this into an epic & chunk it up it the most logical way w/ Subha's feedback

**Comment by SubhaShroff** [comment]
@TracyLinder  @Nadin-Nader 
As I understand this whole story has to go together in PROD. I was wondering, if we can take this story in 3 parts(sub tasks) until TEST. Once, it is completely UATed in TEST and accepted we can deployed in PROD as a whole.

**Comment by SubhaShroff** [comment]
@vschandramouli  @Nadin-Nader I have couple of questions:
1. noreply@cityofaustin.org - I assume this is going to be FROM email for the emails that to be sent.
2. When we look for people in EX, DS, the people type should be Right of Way contractor just as in LC? Or can be any people type?

**Comment by TracyLinder** [comment]
### Convo w/ ROW During Friday Review 6/9/23


Email To:
Send email to the folderpeople type Right of Way Contractor on the LC folder
From: [noreply@cityofaustin.org](mailto:noreply@cityofaustin.org)
-

**Comment by SubhaShroff** [comment]
@Nadin-Nader  SCTASK0327498 is raised for DEV deployment. Thanks.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  SCTASK0329434 2nd DEV development is submitted. Thanks.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  SCTASK0330807 is submitted to revert  7fe2c6c70860 (SCTASK0329434) on DEV. Need to make a new commit. Once DEV is up will work on that.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  SCTASK0331137 is submitted for another commit's DEV deployment.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  The commits are deployed in DEV. I tested for the following:
1. New Batch to update the EX, DS permits of Right Of Way contractor associated with an expired LC: EX, DS permits are updated with 'Expired - License' status. -- Sample RSNs - 12498993(EX), 12501966(EX), 12464954(DS). 
2. Expire EX - Batch job is missing in DEV. Will recreate it from DEV and test. - TESTING..........................
3. Expire DS - Batch job(47 in DEV - Expire BMEP Pemits whose expiry data has passed) - TESTING..........................
4. Re-activation process Acceptance Criteria -  YET TO TEST......

**Comment by SubhaShroff** [comment]
@Nadin-Nader  SCTASK0333359 submitted for TEST deployment.

**Comment by TracyLinder** [comment]
### Friday ROW Review

- Subha needs to understand the user workflow of the LC folder so she can do proper testing. 
- Kyle confirmed the process 'Renew License' (58025) is being used
- Is there an existing training document for this? --- Kyle says yes, it might be useful. @KyleErich please drop this doc in the comments

**Comment by Nadin-Nader** [comment]
Still Waiting on CTM to deploy to TEST

**Comment by SubhaShroff** [comment]
@Nadin-Nader  TEST deployment request SCTASK0333359 is still open.  We need to follow up with CTM.
But  SCTASK0331137 is completed. In this request I have mentioned, this has dependency on SCTASK0333359. I will  follow up Martin on Monday.
Thanks.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  - SCTASK0331137 had dependency on SCTASK0333359. Since the deployment order changed, SCTASK0331137 needs redeployment. Will talk to Martin about it.

**Comment by SubhaShroff** [comment]
@Nadin-Nader 
I tested this ticket. 
**1st - Expiration Batch Acceptance Criteria**
Batch job - 196
few of permits that got updated - 10793473
10793473	DS	50032
10556971	DS	50032
12440074	DS	50032
12446310	DS	50032
12450041	EX	50032
12456194	EX	50032
12452498	EX	50032
12451782	EX	50032
12464900	EX	50032
12465672	DS	50032
12475638	DS	50032
12456199	EX	50032
12469607	DS	50032
Emails sent --
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/fa6889fb-fb71-4a9e-bf91-70ef30c2f26d)

Batch Job for EX Expiry in TEST - 179 - RUNNING...........................

Batch Job for DS expiry in TEST - 38 - RUNNING.............................

**Comment by SubhaShroff** [comment]
@Nadin-Nader 
### 3rd - Modification to DS&EX expiration batches
Batch Job for EX Expiry in TEST - 179  - Testing again on TEST after Martin redeployed -- **RUNNING**.. (as I see the previous instances for this job it takes approx half an hour to Complete. Once completed will update here.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  After above investigation, I updated existing procedure Process_BP_00000093 to bypass the Invalid Attempt error for 'Expired - License' status on DS.
SCTASK0339714 is submitted for DEV/TEST deployment of the related commit. - [2a3dcd35cbe7](http://coaamandadevops02.austintexas.gov/R5:2a3dcd35cbe761bfe2e5e114c18d70859847e50f).

**Comment by SubhaShroff** [comment]
Hi Nadin, I tested it and this time it gave ExpireBMEP batch Job gave another issue. This batch job calls so many inspections dynamically. At this point, I am thinking if its better to create a new batch job to expire DS with 'Expiry - License' alone. But we need to discuss this.
Thanks.

**Comment by SubhaShroff** [comment]
NOTES after call with Kyle:
for expiring DS, create a new batch job. Not use 38 job.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  [d0ca368375f8](http://coaamandadevops02.austintexas.gov/R5:d0ca368375f8252a998b9b37a58156975e2b7864), [9e6ed746b19f](http://coaamandadevops02.austintexas.gov/R5:9e6ed746b19fb9aa74079aeacd385ca50b9f54e7) commits are done to revert the changes done on BMEP related batch job and created a new batch job to expire DS with Expired License status.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  Thank you. For the above commits DEV promotion request is submitted.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  [SCTASK0345041](https://atx.servicenowservices.com/sc_task.do?sys_id=fc56b7651b6839d0a3140fe4604bcbf6&sysparm_record_target=sc_task&sysparm_record_row=1&sysparm_record_rows=1&sysparm_record_list=request_item%3D9856b7651b6839d0a3140fe4604bcb7b%5EORDERBYnumber) is deployed. I raised a request for TEST Manual deployment of Batch Job - SCTASK0345700.

**Comment by SubhaShroff** [comment]
I have requested Anu to serve SCTASK0345700.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  Anu has deployed the Batch Job for expiring DS in TEST. I tested it and looks working. Said that now the entire ticket is ready for UAT. Thanks.

**Comment by TracyLinder** [comment]
### ROW Review
- [ ] Update the LC expiry date fields - map them correctly on both emails
- [ ] User should only get one email

**Comment by SubhaShroff** [comment]
@Nadin-Nader  @TracyLinder 
### LC permit number in email contents:
committing the code..... will update it here soon with commit/servicenow ticket number.

### Duplicate Emails:
Looked into it and the emails sent were for different expired LCs of same Right of Way Contractor. Since the email contents I added DS/EX permit numbers instead of LC permit numbers, it looked like those are duplicates. But those are sent for different LCs.  Please refer to the following PROD data for a similar scenario.
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/d9e4a539-6b9c-4c8a-b286-8f85a191bfeb)

I observed another scenario where for a ROW contractor, there can be one LC, and different DS/EX permits. In this case also, he/she will get multiple emails.
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/8ff5eb69-396d-484b-8beb-ed41c22178f1)

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has been sent to the DevOps CM Team to be deployed to DEV under **SCTASK0349478**

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has been deployed to dev, please  UAT when you get back and when you have a chance.

**Comment by Nadin-Nader** [comment]
Update: UATing on DEV

**Comment by SubhaShroff** [comment]
@Nadin-Nader  I couldn't continue testing due to a compilation error on DEV. I reached Martin yesterday, he looked into it and replied this morning that issue caused by procedure GIS_PULL_REQUEST. He is going to report to GIS team so that they can look into it.
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/e691cd46-c1e5-4091-9a3b-ff8d2b2df712)

**Comment by Nadin-Nader** [comment]
Approved for TEST by ROW

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has been sent to the DevOps CM Team to be deployed to TEST **SCTASK0354619**

**Comment by SubhaShroff** [comment]
@Nadin-Nader  This is deployed in TEST and i tested it. Ready for UAT.

**Comment by Nadin-Nader** [comment]
Sent Kyle a message that this is ready for UAT on TEST. Waiting on Response

**Comment by KyleErich** [comment]
@Nadin-Nader UAT approved on TEST.

**Comment by SubhaShroff** [comment]
Thank you @KyleErich.

@Nadin-Nader  Here is the PROD deployment plan. Please submit promotion request. Thank you.
[10849 PROD Deployment Plan.docx](https://github.com/cityofaustin/atd-data-tech/files/12753079/10849.PROD.Deployment.Plan.docx)

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has been sent to the DevOps CM Team to be deployed to PROD under **SCTASK0359272**

**Comment by SubhaShroff** [comment]
@Nadin-Nader  commits are on PROD. Now manual deployment needs to be done on PROD. Will send a request first thing tomorrow with the deployment plan. Thanks.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  Please raise a PROD manual deployment request.
[10849- PROD Manual Deployment Plan of Batch Job.docx](https://github.com/cityofaustin/atd-data-tech/files/12810086/10849-.PROD.Manual.Deployment.Plan.of.Batch.Job.docx)

**Comment by Nadin-Nader** [comment]
This has been sent to the CTM team to be deployed to PROD under **SCTASK0360887**

**Comment by Nadin-Nader** [comment]
Asked Kyle to UAT this on PROD. Waiting on Kyles response

**Comment by Nadin-Nader** [comment]
Per Kyle:

> Standby on this. I have a license set to expire tomorrow so I'll be able to approve/reject the UAT at that time once I see what happens.

**Comment by Nadin-Nader** [comment]
@SubhaShroff this failed UAT Per Kyle:

> UAT failed unfortunately. I had LC license RowID 13222392 set to expire on 10/16, associated with ROW Contractor RowID 5835282, which was in turn associated with test permits 13109577, 13033817, 12676544, and 12867855 and unfortunately none of these permits went into Expired - License status nor did any automatic email get sent out.

**Comment by Nadin-Nader** [comment]
@SubhaShroff I sent this to CTM to be deployed to PROD under **SCTASK0365201**

**Comment by Nadin-Nader** [comment]
Waiting on Kyles Response

**Comment by Nadin-Nader** [comment]
Update: Need to turn off Batch as it is expiring permits that should not be

**Comment by TracyLinder** [comment]
### Update - PROD BUGS ❌

Per ROW EX & DS folders are updating to Expired - License when they shouldn't be. We will turn off these batch jobs and begin fixing. 

There are currently 300+ folders in this status and ROW will have to review them independently for valid status. 👎🏼

**Comment by Nadin-Nader** [comment]
Turned of both Batch Jobs. Batch Job RSNs are below:

Expire EX DS with LC -- 203

Expire DS with Expire License Status --200

**Comment by SubhaShroff** [comment]
@TracyLinder  @Nadin-Nader I looked at the following example that Kim had pulled in the Friday 2pm call.
### EX-Rsn - 12596263 on PROD got updated to EXPIRED-LICENSE status though Right of Contractor's( MASTEC AUSTIN, TEXAS DIVISION - 5007461),  LC is active.

When I run the query for this EX, I got the following data. It is getting updated because of another Right of Way contractor(ROW)'s (31845) expired LC and that too there are multiple expired LCs for this 31845 ROW. This ROW is a different people type in EX. 
So, looks like the system should check  the people should be ROW in the permits as well before updating those. If that is correct, then I think we don't need to do 24hours check.  **_Please confirm._** 

**EX/DS Rsn                ROW        LC rsn       LC Status**
12596263	EX	31845	1079706	Expired
12596263	EX	31845	1078408	Expired
12596263	EX	31845	1077863	Expired
12596263	EX	31845	1077767	Expired

**Comment by SubhaShroff** [comment]
@Nadin-Nader Updated the batch process and did a commit - [6712bceea373](http://coaamandadevops02.austintexas.gov/R5:6712bceea3730d2d739fdda2d756e918e7c4002a)
Based on Kyle's confirmation, included the condition-LC Right of Way Contractor(ROW) should be ROW in DS, EX also, before updating the status.
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/17b52a90-f9b3-465b-958d-fe95f0e0e73c)

Here is the deployment plan. Please deploy it. - 
[10849- 6712bceea373 - DEV TEST Deployment Plan.docx](https://github.com/cityofaustin/atd-data-tech/files/13243675/10849-.6712bceea373.-.DEV.TEST.Deployment.Plan.docx)

**Comment by Nadin-Nader** [comment]
This was sent to the DevOps CM Team to be deployed to DEV and TEST under **SCTASK0368623**

**Comment by SubhaShroff** [comment]
@Nadin-Nader The corrected batch job Expire DS EX with LC (198) is on DEV. I tested it. Would like to test it with Vishwa or you, before sending it TEST.

**Comment by Nadin-Nader** [comment]
UAT scheduled for 11/9 @11am

**Comment by SubhaShroff** [comment]
Batch job on TEST:
Expire EX DS with LC - 196
Expire DS with Expire License Status - 200
Expire EX Folders - 179

**Comment by TracyLinder** [comment]
### 11/9/23 UAT
Tracy, Vish, Subha, Nate, Kyle

Takeaways:
- [ ] system should send one email to the license holder with a list of all expired permits. Currently it is sending an email for each permit. 
- [ ] Figure out why the URLs look so bad


![image](https://github.com/cityofaustin/atd-data-tech/assets/49930868/b78d5d26-e0f8-4a10-9054-4a5a32fcda9d)

**Comment by SubhaShroff** [comment]
@Nadin-Nader  @TracyLinder 

- URLs - as we spoke in a call, the URLs look messy only in austintexas.gov inboxes. Looks like the URLs are wrapped with safe links which is not worrisome and by clicking those will still work as expected.
- Batch emails are fixed the multiple emails to single email for 
-  I have fixed the multiple emails to single email for Renew process also.

The code changes are committed - 01539992caa191f37d71ab8bd18bea8cdbf42294. @Nadin-Nader  Please raise a DEV promotion request. Thanks.

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has been sent to the DevOps CM Team to be deployed to TEST under **SCTASK0372166**

**Comment by Nadin-Nader** [comment]
This is ready for our UAT on DEV

**Comment by Nadin-Nader** [comment]
@SubhaShroff the new changes have been sent to the DevOps CM Team to be deployed to DEV under **SCTASK0374131**

**Comment by SubhaShroff** [comment]
@SubhaShroff I tested it on DEV and it looks working as expected with the emails. Please raise a TEST deployment for this. Thank you.
[10849-01539992caa191f3- 9431541f640830 - TEST Deployment Plan.docx](https://github.com/cityofaustin/atd-data-tech/files/13507824/10849-01539992caa191f3-.9431541f640830.-.TEST.Deployment.Plan.docx)

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has been sent to the DevOps CM Team to be deployed to TEST under **SCTASK0374881**

**Comment by Nadin-Nader** [comment]
@SubhaShroff please UAT on TEST

**Comment by SubhaShroff** [comment]
@Nadin-Nader  I tested this ticket after working on emails to send one email to each right of way contractor. This is ready for UAT on TEST. Thanks.

**Comment by TracyLinder** [comment]
### UAT Cases

Contractor LC: True Dynamic Site Solutions, LLC (Michael W Rood)

EX: 13043732 
DS:

**Comment by TracyLinder** [comment]
@vschandramouli can you please schedule a facilitated session for this?

**Comment by KyleErich** [comment]
@TracyLinder @vschandramouli @SubhaShroff I need to reject the UAT for this and I'll explain why. I think the same problem that caused us to roll this back from AMANDA PROD is continuing.

When you query the Expired - License EX & DS folders in AMANDA TEST, you get 249 results. Some of these correctly went into Expired - License, including the ones we have been UAT'ing, because the LC folder associated with their ROW Contractor went into Expired status. HOWEVER, it is also putting EX & DS folders into Expired - License status even when there is an Active LC associated with the ROW Contractor, and it's doing this because that ROW Contractor record is associated with multiple LC records -- including at least one that is Expired. For example, if a ROW Contractor is associated with one Active LC record, but also two Expired LC records, then its permits will go into Expired - License even though there is an Active LC record because the two Expired LC records are taking precedence.

An example in AMANDA TEST is ROW Contractor RowID 79938. Its associated EX permit RowID 13184541 went into Expired - License. When you query this ROW Contractor People ID 79938 in the License - Professional module, you get two LC records, one Active and one Expired. LC RowID 10101666 is Active because Insurance Expiration and Bond Expiration dates are still current; however, RowID 1079298 is Expired because it is a disused migrated record from the past SUEP system with very old expiration dates but still associated with the ROW Contractor.

**We need some way for the system to determine that if there is at least one Active LC associated with a ROW Contractor, then it should NOT send the command to expire all the permits associated with this ROW Contractor record.**

If this is not possible, I think we may need to ponder other ways of detaching duplicate LC records from ROW Contractor records, like running some batch job to remove any records containing the term "SUEP" in their Folder name and/or any other parameters relevant to disused LC records that can isolate them from their associated ROW Contractor.

**Comment by SubhaShroff** [comment]
Thanks for the feedback Kyle.
@Nadin-Nader  I did the changes to address Kyle's feedback and did a commit - 554ca0d262630ec8a5786a342c94e18adbf3a35d.
@Lpriya-LIT  please create a DEV deployment request.
[10849- 554ca0d262630ec8a5786a342c94e18adbf3a35d- DEV Deployment Plan.docx](https://github.com/cityofaustin/atd-data-tech/files/13773577/10849-.554ca0d262630ec8a5786a342c94e18adbf3a35d-.DEV.Deployment.Plan.docx)

**Comment by Lpriya-LIT** [comment]
@SubhaShroff Created Dev deployment request , SCTASK0381158

**Comment by SubhaShroff** [comment]
Thanks @Lpriya-LIT. 
I checked SCTASK0381158, it looks closed. But I don't see the code deployed. I  will follow up with the CTM person.

**Comment by SubhaShroff** [comment]
@Lpriya-LIT Priya please create a DEV deployment request.
[10849- d94973b7024d34f489e307383913bc660b329f42- DEV Deployment Plan.docx](https://github.com/cityofaustin/atd-data-tech/files/13782052/10849-.d94973b7024d34f489e307383913bc660b329f42-.DEV.Deployment.Plan.docx)

**Comment by SubhaShroff** [comment]
@Nadin-Nader  @Lpriya-LIT  SCTASK0381158 is reverted on DEV due to conflicts and not properly deployed. CTM person Monty tried to correct it but he said he would take George's help for this. I see George is on holiday until Jan 2nd. We will get it sorted that on or after that day.

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has been sent to the DevOps CM Team to be deployed to DEV under **SCTASK0382558**

**Comment by Nadin-Nader** [comment]
@SubhaShroff This has been sent to the DevOps CM Team to be deployed to TEST under **SCTASK0383515**

**Comment by TracyLinder** [comment]
@KyleErich will help us setup test data to run this batch job. We will need the following:

1. LC, Active with expiration date of 1/10/24
2. Connect 2 permits: Ex and DS permit, active, expiration date after 1/10 connected to the LC record
4. Connect 2 permits: EX and DS permit, active expiration date 1/10 connected to the LC record
5. Connect 2 permits: EX and DS, expired with expiration in the past

**Comment by Nadin-Nader** [comment]
@vschandramouli @SubhaShroff   This has been deployed to TEST and in order to UAT we will still need to do what Tracy mentioned in the previous comment

**Comment by KyleErich** [comment]
AMANDA Test RowIDs for the requested scenarios:

1. 13204187
2. 13110678; 13203188
3. 13189434; 13142614
4. 13082746; 13074647

Associated License RowID: 13204187
Associated ROW Contractor PeopleID: 5831071

I also associated an expired license with the ROW Contractor so that we can also test to make sure the active associated license overrides the expired one. That RowID is 13201304.

**Comment by KyleErich** [comment]
@SubhaShroff 

Hyperlinks for "ROW License and Permit Expiration Notice" email:

All Driveway/Sidewalk and Excavation permits associated with your Right of Way Contractor License have expired on 10-JAN-24 until the license is renewed. A list of those permits is provided below. To renew your Right of Way Contractor License, see the ''How is a Right of Way Contractor License renewed?'' section [at this link](https://www.austintexas.gov/page/right-way-contractors).

To update your insurance, bond or license, please first see the documentation requirements on the [Right of Way Contractor webpage](https://www.austintexas.gov/page/right-way-contractors), and then submit your updated documentation to [the linked Smartsheet form](https://app.smartsheet.com/b/form/cad1e53c88224399979f40df80ef6270).

**Comment by SubhaShroff** [comment]
@Nadin-Nader As per Kyle's suggestion to update the URLs to Hyperlink I did the necessary changes and committed the code.
d7a96954d32e91c64f4e91961fb53bd5f3a5b966
42ac71795c84d5b353e4e96f398e1dd64ffbe247
Here is the DEV deployment request submitted for the above two commits - SCTASK0386003.

**Comment by SubhaShroff** [comment]
@Nadin-Nader I tested the hyperlink changes in the email content on DEV.
Priya submitted a TEST deployment request for the same - SCTASK0386568.
Thanks.

**Comment by SubhaShroff** [comment]
@Nadin-Nader  @KyleErich Expiration Notice emails contents are updated to the suggested Hyperlinks in TEST. I tested this and the content looks as follows. Please review and let me know if this ticket can be promoted to PROD.
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/29a85ef7-dbf7-47f9-af2a-d4c6fb6c8cb2)

**Comment by KyleErich** [comment]
@SubhaShroff @Nadin-Nader I did a test and it looks like I'm still receiving the older version of the email, with the hyperlink issues and the issues with line breaks:
![image](https://github.com/cityofaustin/atd-data-tech/assets/130776506/f5ae1fcf-e625-4415-98e7-0e3caf8374c6)

**Comment by SubhaShroff** [comment]
Sorry about that @KyleErich. 
@Nadin-Nader I checked for this, the code is intact but I saw VALIDLOOKUP table changes were changed. I spoke to George this morning and he re-applied the respective commit in TEST. I tested it now and the email looked good.
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/97279b0b-11fd-4470-be03-5cfb3deefcdd)
Please UAT.
Thanks.

**Comment by KyleErich** [comment]
On my second test in TEST, the license (13204187) expired but the associated permits (13203188 and 13110678) remain Active. FYI.

**Comment by SubhaShroff** [comment]
@KyleErich As per AC, the batch should check the LC that expires in last 24 hours. So, LC(13204187)  expired today, the associated EX/DS will get updated tomorrow and an email will be sent with the batch trigger around 12.05 AM. Please look for the results tomorrow for this LC.

**Comment by KyleErich** [comment]
@SubhaShroff Will do, thank you.

**Comment by KyleErich** [comment]
@SubhaShroff @Nadin-Nader UAT on Test approved. Thanks for your hard work on this card, I know this one has been a thorn in our side for a long time. Ready to promote to Prod.

**Comment by Nadin-Nader** [comment]
@SubhaShroff this has been sent to the DevOps CM Team to be deployed to PROD under **SCTASK0393770**

**Comment by Nadin-Nader** [comment]
@SubhaShroff the request for the manual batch job to be activated has been sent to the CTM Team to be done on PROD under **SCTASK0394348**

**Comment by Nadin-Nader** [comment]
Still waiting on CTM Team to activate on PROD

**Comment by SubhaShroff** [comment]
@Nadin-Nader @KyleErich  Since the following batches are activated, now this ticket is completely deployed in PROD. Ready for UAT. Thank you.
Job RSN: 203
Job Name:   Expire EX DS with LC

Job RSN: 200
Job Name:   Expire DS with Expire License Status

**Comment by SubhaShroff** [comment]
@Nadin-Nader  @KyleErich I checked with Anu about this. Thats true, the batches didn't trigger, even though the batches are set to on, which is unusual. For today, Anu, fired them manually. Meanwhile I investigated about this and got to know from one of my colleagues that there are some cases with Amanda 7. 
The suggested solution just to update batch job NextFireDate and set it back to future date. So, since Anu already ran those batches and the date got updated to 2-17-2024. So that should give a push for auto-trigger these batches. Will check for this on Monday.

**Comment by SubhaShroff** [comment]
@Nadin-Nader @KyleErich 
1. As expected I see the batches are automatically triggering.
2. Today the following LC are expiring. So tomorrow the respective Right-of-Way contractors must be getting emails about it.
ROW contr         LC rsn
----------------------- 
5660107	12142442
5585672	11700670
5837007	13233278
5384754	11029120
Out of the above listed Right of Way Contractors, one has two DS permits of (Right of Way Contractor: 5660107)should be getting updated to Expired-License.
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/660f7737-58d5-4d9d-9203-0a71d0b7ab7b)

Thank you.

**Comment by KyleErich** [comment]
@SubhaShroff Thank you for looking into this. Is it possible that the batches are only automatically triggering for the DS folders and not for the EX folders? Because the example I gave earlier associated with LC RSN 12136760 is still showing Active EX permits, despite the LC being expired.

A query of all "Expired - License" folders is only showing DS folders currently:
![image](https://github.com/cityofaustin/atd-data-tech/assets/130776506/2fb14d24-42fd-4cbf-b36c-455419a02eba)

**Comment by SubhaShroff** [comment]
@KyleErich  As per the AC, the batch job should pick only the LC folders that expired the previous day(in last 24 hours when the batch job runs). The LC that you have mentioned above has expired around 12th Feb.
Following are the LCs that expired yesterday.
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/d51c04bb-0d50-4e77-8c35-71f31b7132fa)

The respective DS folders that got updated to 'Expired-License' and stayed so, because their expiry date is future date.
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/d812bd71-1294-486e-a8c5-3ae52c1d1ef3)

Few of EX/DS folders of both LCs further got updated to 'Expired' status as those have past date(with EX/DS expire batch jobs triggered), listed below:
![image](https://github.com/cityofaustin/atd-data-tech/assets/51680715/427443ab-7225-4566-8baf-b472cdb7ba5e)

**Comment by TracyLinder** [comment]
This isn't quite stable on PROD however we've opened other tickets to address the bugs. Closing.
